lib/install.js

```diff
@@ -544,11 +544,20 @@ Installer.prototype.executeActions = function (cb) {
 
 Installer.prototype.rollbackFailedOptional = function (staging, actionsToRun, cb) {
   if (!this.rollback) return cb()
+  // get list of all failed optional installs, potentially with duplicates
   var failed = actionsToRun.map(function (action) {
     return action[1]
   }).filter(function (pkg) {
     return pkg.failed && pkg.rollback
   })
+  // remove duplicates to avoid concurrent rollbacks of the same package
+  failed = failed.filter(function (pkg, i) {
+    return failed.indexOf(pkg) === i
+  })
+  // also remove child packages since they are rolled back by their parents
```

Comment:
Actions should be sorted deepest first so strictly speaking  this shouldn't be necessary. This would also fail to prune deeply nested failed optionals (eg one's where `pkg.parent.parent` is failing and `pkg` is failing but `pkg.parent` is not).

Reply:
Yes, actions are sorted deepest first. But they are executed asynchronously. Therefore the delete operation on a top level package may start when the delete operation on a subpackage is still going on.